src/services/widgets/widgets.svc.coffee

```diff
@@ -115,102 +111,114 @@ angular
             for widget in currentDhb.widgets
               newMetadata = angular.merge({}, widget.metadata, metadata)
               # If the metadata has not changed, we don't push the promise
-              unless _.isEqual(widget.metadata, newMetadata)
-                widget.isLoading = true
-                promises.push _self.update(widget, {metadata: newMetadata})
+              promises.push _self.update(widget, {metadata: newMetadata}, false) unless _.isEqual(widget.metadata, newMetadata)
 
             $q.all(promises).then( (results) -> _self.refreshAll(refreshCache) )
 
       )
 
     @isRefreshing = false
     @refreshAll = (refreshCache=false) ->
-      unless _self.isRefreshing
-        _self.isRefreshing = true
-        _self.load().then( ->
+      return $q.resolve() if _self.isRefreshing
+
+      _self.isRefreshing = true
+      _self.load().then(
+        ->
           currentDhb = ImpacDashboardsSvc.getCurrentDashboard()
+          promises = []
           for w in currentDhb.widgets
-            w.isLoading = true
-            _self.show(w, refreshCache).then(
-              (renderedWidget) -> renderedWidget.isLoading = false
+            promises.push _self.show(w, refreshCache).then(
+              (renderedWidget) ->
+                $q.resolve(renderedWidget)
+              (errorResponse) ->
               # TODO: better error management
-              (errorResponse) -> $log.error(errorResponse.data.error) if (errorResponse.data? && errorResponse.data.error)
+                $log.error(errorResponse.data.error) if (errorResponse.data? && errorResponse.data.error)
+                $q.reject(errorResponse)
             )
-        ).finally(->
+          $q.all(promises)
+      ).finally(
+        ->
           # throttles refreshAll calls (temporary fix until rx.angular.js is implemented)
-          $timeout(->
-            _self.isRefreshing = false
-          , 3000)
-        )
+          $timeout(
+            ->
+              _self.isRefreshing = false
+            3000
+          )
+      )
 
     # ====================================
     # CRUD methods
     # ====================================
 
-    @show = (widget, refreshCache=false) ->
-      deferred = $q.defer()
-
+    initWidget = (widget) ->
+      # Init
+      widget.initContext() if angular.isDefined(widget.initContext)
+      # Init settings
+      unless _.isEmpty(widget.settings)
+        for setting in widget.settings
+          setting.initialize() if angular.isDefined(setting.initialize)
+      # Draw chart
+      widget.format() if angular.isDefined(widget.format)
+
+    # Calls Legacy Impac! or bolt API to render a widget
+    # If no content is returned, the endpoint will be called again with `demo` = `true` to retrieve stub data
+    @show = (widget, refreshCache = false, demo = false) ->
```

Comment:
Maybe the demo flag should be within an `opts` arg instead? More flexible with future additions.